### ISSUE TITLE
Explicitly implement some members in System.Collections.Specialized?

### ISSUE THREAD

**Issue by justinvp** [dotnet/corefx#370]
Should some of the implicitly implemented members in `System.Collections.Specialized` be implemented explicitly? Is it too late to do this?

The following members are explicitly implemented (and therefore public to intellisense):

```c#
public class HybridDictionary : IDictionary
{
    public bool IsReadOnly { get; }
    public bool IsFixedSize { get; }
    public bool IsSynchronized { get; }
    public object SyncRoot { get; }
    public void CopyTo(Array array, int index);
    
    ...
}

public class ListDictionary : IDictionary
{
    public bool IsReadOnly { get; }
    public bool IsFixedSize { get; }
    public bool IsSynchronized { get; }
    public object SyncRoot { get; }
    public void CopyTo(Array array, int index);

    ...
}

public class NameValueCollection : NameObjectCollectionBase
{
    public void CopyTo(Array dest, int index);
    
    ...
}

public class StringCollection : IList
{
    public bool IsReadOnly { get; }
    public bool IsSynchronized { get; }
    public object SyncRoot { get; }
    
    ...
}

public class StringDictionary : IEnumerable
{
    // Do these need to be virtual?
    public virtual bool IsSynchronized { get; }
    public virtual object SyncRoot { get; }
    public virtual void CopyTo(Array array, int index);

    ...
}
```

## Background

When the `Regex` collections were subsetted and added here (to corefx), some of the members were deliberately changed from implicit to explicit.

For example, [in the full .NET framework](https://github.com/Microsoft/referencesource/blob/9da503f9ef21e8d1f2905c78d4e3e5cbb3d6f85a/System/regex/system/text/regularexpressions/RegexCaptureCollection.cs#L41-L75), `CaptureCollection` has the following members public:

```c#
public class CaptureCollection : ICollection
{
    public Object SyncRoot { get; }
    public bool IsSynchronized { get; }
    public bool IsReadOnly { get; }
    public void CopyTo(Array array, int arrayIndex);
    ...
}
```

Compared to what's in [corefx](https://github.com/dotnet/corefx/blob/cc07fbf1a2fdac12523573ffb2dd273d82542f8d/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCaptureCollection.cs#L32-L49):

```c#
public class CaptureCollection : ICollection
{
    Object ICollection.SyncRoot { get; }
    bool ICollection.IsSynchronized { get; }
    // IsReadOnly removed
    void ICollection.CopyTo(Array array, int arrayIndex);
    ...
}
```

**Comment by davkean** [comment]
I'm conflicted on this one. Justin, you and I are in agreement on this with regards to libraries and code that we do not consider legacy_ (given you and I were leading making these explict back on the BCL team many years ago) - let's go ahead and hide these somewhat useless properties/methods that forced by the implementation of ICollection for new and relevant code.

However, these collections, along with System.Collections.NonGeneric are entirely there for backwards compatibility - we don't want to invest in them, we don't want to add features to them. There are better and more appropriate replacements for them. Adding features, including generic interfaces, makes them more attractive to use, which is entirely not want we want and on top of that, hiding these public properties/methods, can only break code that was using them (whether correctly, or incorrectly) that ports to .NET Core. 

This also applies to implementing generic interfaces, we know for a fact that this will break serialization on .NET Framework - we recently rolled back our implementation of IDictionary<TKey, TValue> on StringDictionary due to this.

To summarize, I believe that these should be ported as is, warts and all, to .NET Core with very little changes, solely for source compatibility. Let's spend the investment elsewhere, on the libraries and types that we want to invest and bring forward.

**Comment by justinvp** [comment]
Fair enough.

**Comment by akoeplinger** [comment]
@davkean Can we get some `[Obsolete]` applied to the collections considered *legacy*?

One thing I often heard from people new to .NET was that they find it difficult to know which parts of the BCL are considered "new and go forward" and which are there solely for not breaking existing code but with better alternatives available.

**Comment by sharwell** [comment]
@davkean What would you consider the best reference information regarding the impact changes like this have on serialization?

**Comment by pdelvo** [comment]
@akoeplinger I was wondering that too. Why aren't ArrayList and co marked as obsolete?

**Comment by ellismg** [comment]
(Note that this statement is colored by my work on the full .NET Framework).  In the past, we've been hesitant to mark items as obsolete, because in some cases it can be considered a source breaking change (someone is using an obsolete method and building with warnings as errors).

This problem was magnified in cases where we did not have a clean separation between contract assemblies and implementation assemblies and we did in-place updates.

Today, I think we have more leeway in actually obsoleting things since we have a better build time contract story.

**Comment by weshaggard** [comment]
Also in the world of the full .NET framework these source breaking changes can turn into runtime breaking changes for dynamic compilation scenarios like ASP.NET compiling code on the server. So in a world where we update the libraries in place marking something obsolete is actually a breaking change, however the more we move libraries to be app local I think that risk goes down and the developers are the ones making the updates and the choice to ignore the obsolete APIs or to fix them. We should definitely start considering this for .NET Core.

**Comment by terrajobst** [comment]
I agree wth @weshaggard on the risk but I'm not sure that marking these types as obsolete is the right approach. If somebody uses `ArrayList` they are probably doing it of for legacy reasons. It doesn't seem to help if compiling against the .NET Core version causes hundreds of warnings...

**Comment by akoeplinger** [comment]
@terrajobst those warnings can be easily disabled, I don't think that's a good enough reason to keep confusing users.

I understand why you couldn't do it in the past but if this framework is going to last for the next decade, guiding users to the right APIs via applying `[Obsolete]` to the APIs everyone on your team considers legacy sounds like a valid approach to me :)

**Comment by sharwell** [comment]
Regarding explicit interface implementation: Considering the *only* reason this assembly gets included is for improving backwards compatibility with certain legacy code, I would be :-1: on any breaking change to the API surface even if the members are "unlikely to be used".

Regarding `[Obsolete]` I wouldn't be opposed to adding this attribute on a case-by-case basis. There was never really any confusion from my end because the lack of implementing generic interfaces was a clear indication that these are not the newest APIs available. Either way, I recommend creating a separate issue for `[Obsolete]` and closing *this* issue.

**Comment by weshaggard** [comment]
@sharwell I also agree we should speared the conversations and I also agree we shouldn't invest in adding these interfaces to the S.C.Specialized or S.C.NonGeneric types. So with that said I'm closing this issue.

**Comment by terrajobst** [comment]
@akoeplinger 

>  Those warnings can be easily disabled, I don't think that's a good enough reason to keep confusing users.

But only at the global level. You can't say "yes, I know about `ArrayList`". These warnings will cause a lot of noise and my concern is that you miss other obsoletion warnings that tell customers about potentially issues in their applications. In the past, we've been quite selective when it comes to applying this attribute. If used to broadly, the signal to noise ration reaches the point where the features becomes useless.
